sockeye/inference.py

```diff
@@ -379,35 +399,46 @@ def load_models(context: mx.context.Context,
     """
-    models, source_vocabs, target_vocabs = [], [], []
+    models, source_vocabs, source_factor_vocabs, target_vocabs = [], [], [], []
     if checkpoints is None:
         checkpoints = [None] * len(model_folders)
+
+    source_factor_vocabs = []
     for model_folder, checkpoint in zip(model_folders, checkpoints):
         source_vocabs.append(vocab.vocab_from_json_or_pickle(os.path.join(model_folder, C.VOCAB_SRC_NAME)))
         target_vocabs.append(vocab.vocab_from_json_or_pickle(os.path.join(model_folder, C.VOCAB_TRG_NAME)))
-        model = InferenceModel(model_folder=model_folder,
-                               context=context,
-                               beam_size=beam_size,
-                               batch_size=batch_size,
-                               softmax_temperature=softmax_temperature,
-                               checkpoint=checkpoint,
-                               decoder_return_logit_inputs=decoder_return_logit_inputs,
-                               cache_output_layer_w_b=cache_output_layer_w_b)
-        models.append(model)
+
+        inf_model = InferenceModel(model_folder=model_folder,
+                                   context=context,
+                                   beam_size=beam_size,
+                                   batch_size=batch_size,
+                                   softmax_temperature=softmax_temperature,
+                                   checkpoint=checkpoint,
+                                   decoder_return_logit_inputs=decoder_return_logit_inputs,
+                                   cache_output_layer_w_b=cache_output_layer_w_b)
+        models.append(inf_model)
+
+        source_factor_vocabs.append(load_source_factor_vocabs(model_folder, inf_model.num_factors))
 
     utils.check_condition(vocab.are_identical(*source_vocabs), "Source vocabulary ids do not match")
     utils.check_condition(vocab.are_identical(*target_vocabs), "Target vocabulary ids do not match")
 
+    # sanity check on source factor vocabs
+    if len(source_factor_vocabs) > 0:
+        for fi in range(len(source_factor_vocabs[0])):
+            utils.check_condition(vocab.are_identical(*[source_factor_vocabs[i][fi] for i in range(len(source_factor_vocabs))]),
+                                  "Source factor vocabs do not match")
+
     # set a common max_output length for all models.
     max_input_len, get_max_output_length = models_max_input_output_length(models,
                                                                           max_output_length_num_stds,
                                                                           max_input_len)
-    for model in models:
-        model.initialize(max_input_len, get_max_output_length)
+    for inference_model in models:
+        inference_model.initialize(max_input_len, get_max_output_length)
 
-    return models, source_vocabs[0], target_vocabs[0]
+    return models, source_vocabs[0], target_vocabs[0], source_factor_vocabs[0]
```

Comment:
can we extend the concept of surface source form just being the primary factor in a variable length list of factors further and say that we return a list of vocabs for the source side, instead of splitting the factor vocabs off?